node/disconnect_queue.go

```diff
@@ -42,27 +42,21 @@ func NewDisconnectQueue(node *Node, rate int) *DisconnectQueue {
 		rate:       rateDuration,
 		log:        ctx,
 		shutdown:   false,
-		shutdownCh: make(chan bool),
+		shutdownCh: make(chan struct{}),
 	}
 }
 
 // Run starts queue
 func (d *DisconnectQueue) Run() {
 	throttle := time.NewTicker(d.rate)
-	defer throttle.Stop()
 
 	for {
 		select {
 		case session := <-d.disconnect:
 			<-throttle.C
 			d.node.DisconnectNow(session)
 		case <-d.shutdownCh:
-			d.mu.Lock()
```

Comment:
These checks look too defensive. Was there a reason for them?

Reply:
There was but I don't remember) Looks like the potential race conditions are already covered in other places.

Don't we need to add `close(d.shutdownCh)`?

Also, why not leaving `defer throttle.Stop()` at the top of the `Run` function (as it was before). Isn't it more idiomatic?)

Reply:
> There was but I don't remember) Looks like the potential race conditions are already covered in other places.

I guess it was your first try before `DisconnectQueue#shutdown` introducing but it hadn't help  'cause there is no way to handle  it in `DisconnectQueue#Enqueue`

> Don't we need to add close(d.shutdownCh)?

Good question! And the answer is NO. Channels are not file descriptor. We only need to close them to notify consumers that there would be no more messages in it. In our case, the consumer has already been stopped.

> Also, why not leaving defer throttle.Stop() at the top of the Run function (as it was before). Isn't it more idiomatic?)

I know that it's not a big deal in that particular case, but it makes sense avoid `defer` overusing since it has significant performance drawbacks. Besides, it hinders readability.

Reply:
> it has significant performance drawbacks

Even if we call this function once?)

> Besides, it hinders readability.

I can't see how _hiding_ cleanup call in the `for ... select` statement improves the readability.

Reply:
> Even if we call this function once?)

That's why I wrote

>> I know that it's not a big deal in that particular case

> I can't see how hiding cleanup call in the for ... select statement improves the readability.

I'll put it back if you're worried. Anyway, it was me who put it there in the first place